src/proxy/ssh/server.js

```diff
@@ -0,0 +1,683 @@
+              } else {
+                console.log(`[SSH] Password authentication failed for user ${ctx.username}`);
+                ctx.reject();
+              }
+            } else {
+              console.log(`[SSH] User ${ctx.username} not found or no password set`);
+              ctx.reject();
+            }
+          } catch (error) {
+            console.error('[SSH] Error during password authentication:', error);
+            ctx.reject();
+          }
+        } else {
+          console.log('[SSH] Password authentication attempted but public key was provided');
+          ctx.reject();
+        }
+      } else {
+        console.log(`Unsupported authentication method: ${ctx.method}`);
+        ctx.reject();
+      }
+    });
+
+    client.on('ready', () => {
+      console.log(`[SSH] Client ready: ${client.username}`);
+      client.on('session', this.handleSession.bind(this));
+    });
+  }
+
+  async handleSession(accept, reject) {
+    const session = accept();
+    session.on('exec', async (accept, reject, info) => {
+      const stream = accept();
+      const command = info.command;
+
+      // Parse Git command
+      console.log('[SSH] Command', command);
+      if (command.startsWith('git-')) {
+        // Extract the repository path from the command
+        // Remove quotes and 'git-' prefix, then trim any leading/trailing slashes
+        const repoPath = command
+          .replace('git-upload-pack', '')
+          .replace('git-receive-pack', '')
+          .replace(/^['"]|['"]$/g, '')
+          .replace(/^\/+|\/+$/g, '');
+
+        const req = {
+          method: command.startsWith('git-upload-pack') ? 'GET' : 'POST',
+          originalUrl: repoPath,
+          isSSH: true,
+          headers: {
+            'user-agent': 'git/2.0.0',
+            'content-type': command.startsWith('git-receive-pack')
+              ? 'application/x-git-receive-pack-request'
+              : undefined,
+          },
+        };
+
+        try {
+          console.log('[SSH] Executing chain', req);
+          const action = await chain.executeChain(req);
+
+          console.log('[SSH] Action', action);
+
+          if (action.error || action.blocked) {
+            // If there's an error or the action is blocked, send the error message
+            console.log(
+              '[SSH] Action error or blocked',
+              action.errorMessage || action.blockedMessage,
+            );
+            stream.write(action.errorMessage || action.blockedMessage);
+            stream.end();
+            return;
+          }
+
+          // Create SSH connection to GitHub using the Client approach
+          const { Client } = require('ssh2');
+          const remoteGitSsh = new Client();
+
+          console.log('[SSH] Creating SSH connection to remote');
+
+          // Get remote host from config
+          const remoteUrl = new URL(config.getProxyUrl());
+
+          // TODO: Connection options could go to config
```

Comment:
Bringing attention to this `TODO`. Is there a reason why the default port is `22` in some places, but `2222` in the sample `proxy.config.json`? 🤔

Reply:
Answer: `22` is the default SSH port for connecting to the Git server, whereas `2222` is an arbitrarily-defined port for the proxy server. We might want to add a constant just to clarify the purposes of each.